libp2p/identify.nim

```diff
@@ -0,0 +1,106 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos
+import protobuf/minprotobuf, peerinfo, 
+       protocol as proto, connection,
+       peer, crypto/crypto, multiaddress
+
+const IdentifyCodec* = "/ipfs/id/1.0.0"
+const IdentifyPushCodec* = "/ipfs/id/push/1.0.0"
+const ProtoVersion* = "ipfs/0.1.0"
+const AgentVersion* = "nim-libp2p/0.0.1"
+
+#TODO: implment push identify, leaving out for now as it is not essential
+
+type
+  # TODO: we're doing protobuf manualy, this is only temporary
+  ProtoField[T] = object
+    index: int
+    field: T
+
+  IdentifyInfo* = object
+    pubKey*: PublicKey
+    addrs*: seq[MultiAddress]
+    observedAddr*: MultiAddress
+    protoVersion*: string
+    agentVersion*: string
+
+  Identify* = ref object of LPProtocol
+
+proc encodeMsg*(peerInfo: PeerInfo, observedAddrs: Multiaddress): ProtoBuffer = 
+  result = initProtoBuffer()
+
+  result.write(initProtoField(1, peerInfo.peerId.publicKey.getBytes()))
+  for ma in peerInfo.addrs:
+    result.write(initProtoField(2, ma.data.buffer))
+
+  for item in peerInfo.protocols:
+    result.write(initProtoField(3, item))
+
+  result.write(initProtoField(4, observedAddrs.data.buffer))
+
+  let protoVersion = ProtoVersion
+  result.write(initProtoField(5, protoVersion))
+
+  let agentVersion = AgentVersion
+  result.write(initProtoField(6, agentVersion))
+  result.finish()
+
+proc decodeMsg*(buf: seq[byte]): IdentifyInfo = 
+  var pb = initProtoBuffer(buf)
+
+  var pubKey: PublicKey
+  if pb.getValue(1, pubKey) > -1:
+    result.pubKey = pubKey
+
+  result.addrs = newSeq[MultiAddress]()
+  var address = newSeq[byte]()
+  while pb.getBytes(2, address) != -1:
```

Comment:
```Nim
while pb.getBytes(2, address) > 0:
  result.addrs.add(MultiAddress.init(address))
```
should achieve the same result? no?

Reply:
I'm actually not sure, I stripped it almost verbatium from https://github.com/status-im/nim-libp2p/blob/master/libp2p/daemon/daemonapi.nim#L825-L835, is there any reason for this temps and copies @cheatfate ?

Reply:
Is there any chances the list of multiaddresses contains an empty/zero length in the middle of the list? e.g ["address a", "address b", "address c", "", "address e"] <- the 'd' is empty/zero length.

if 'Yes' then my suggestion won't work. if 'No' then there is a chance we can simplify the code.

Reply:
I mean the multiaddress list encoded in protobuf.

Reply:
It shouldn't - there shouldn't be any gaps in the address list, but the spec is ambiguous about it, so I would expect different implementations handling in different ways. In general, when it comes to networking data, we can't make any assumptions about it. There are going to be as many flavors as there are implementations.

Reply:
true, the spec is too short when describing the listen addresses. I agree we should play safe here.